### PR TITLE
point-cloud 3D GUI

### PR DESCRIPTION
**Pull request by NEDJIMAbelgacem**
## Description
Adds initial GUI for 3D point cloud configuration. For now it is possible to only change the point size.
I tried to make it as close as possible to other layers configuration so I added QgsPointCloud3DSymbol class alongside other GUI classes.
For some reason, QgsPointCloud3DSymbol member functions don't get linked into QGIS if I don't make them defined in the header file. I tried to fix the problem but I don't yet know why it is not working. Any help would be really appreciated.
There is also a problem when the user tries to change the size of the point cloud before QGIS finishes loading the point cloud chunks and it crashes. I will try looking into this next.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@NEDJIMAbelgacem there's some overlap here with https://github.com/qgis/QGIS/pull/40019 -- specifically the layer styling and properties dialog work. I suggest we get https://github.com/qgis/QGIS/pull/40019 merged first as the approach is slightly cleaner in that PR, and then rebase this one to remove the duplicate changes.

File: src/3d/qgs3dutils.cpp [hunk 1]
```diff
@@ -36,6 +36,7 @@
 #include "qgsline3dsymbol.h"
 #include "qgspoint3dsymbol.h"
 #include "qgspolygon3dsymbol.h"
+//#include "qgspointcloud3dsymbol.h"
```
Comment (nyalldawson) [inline src/3d/qgs3dutils.cpp:39]: Remove

File: src/3d/qgspointcloudlayer3drenderer.cpp [hunk 2]
```diff
@@ -73,7 +79,12 @@ Qt3DCore::QEntity *QgsPointCloudLayer3DRenderer::createEntity( const Qgs3DMapSet
   if ( !pcl || !pcl->dataProvider() || !pcl->dataProvider()->index() )
     return nullptr;
 
-  return new QgsPointCloudLayerChunkedEntity( pcl->dataProvider()->index(), map );
+  return new QgsPointCloudLayerChunkedEntity( pcl->dataProvider()->index(), map, mSymbol );
```
Comment (nyalldawson) [inline src/3d/qgspointcloudlayer3drenderer.cpp:82]: This should clone the symbol, and have QgsPointCloudLayerChunkedEntity take ownership of the passed symbol

File: src/3d/qgspointcloudlayer3drenderer.cpp [hunk 3]
```diff
@@ -73,7 +79,12 @@ Qt3DCore::QEntity *QgsPointCloudLayer3DRenderer::createEntity( const Qgs3DMapSet
...
+}
+
+void QgsPointCloudLayer3DRenderer::setSymbol( QgsPointCloud3DSymbol *symbol )
+{
+  mSymbol = dynamic_cast<QgsPointCloud3DSymbol *>( symbol->clone() );
```
Comment (nyalldawson) [inline src/3d/qgspointcloudlayer3drenderer.cpp:87]:
```suggestion
  mSymbol.reset( dynamic_cast<QgsPointCloud3DSymbol *>( symbol ) );
```

File: src/3d/qgspointcloudlayer3drenderer.cpp [hunk 4]
```diff
@@ -83,13 +94,23 @@ void QgsPointCloudLayer3DRenderer::writeXml( QDomElement &elem, const QgsReadWri
...
 }
 
 void QgsPointCloudLayer3DRenderer::readXml( const QDomElement &elem, const QgsReadWriteContext &context )
 {
-  Q_UNUSED( context )
-
   mLayerRef = QgsMapLayerRef( elem.attribute( QStringLiteral( "layer" ) ) );
+
+  QDomElement elemSymbol = elem.firstChildElement( QStringLiteral( "symbol" ) );
+  if ( !mSymbol ) mSymbol = new QgsPointCloud3DSymbol;
```
Comment (nyalldawson) [inline src/3d/qgspointcloudlayer3drenderer.cpp:112]:
```suggestion
  if ( !mSymbol )
     mSymbol = qgis::make_unique< QgsPointCloud3DSymbol >();
```

File: src/3d/qgspointcloudlayer3drenderer.h [hunk 5]
```diff
@@ -69,12 +70,18 @@ class _3D_EXPORT QgsPointCloudLayer3DRenderer : public QgsAbstract3DRenderer
...
+    //! Returns 3D symbol associated with the renderer
+    const QgsPointCloud3DSymbol *symbol() const { return mSymbol; }
+
     void writeXml( QDomElement &elem, const QgsReadWriteContext &context ) const override;
     void readXml( const QDomElement &elem, const QgsReadWriteContext &context ) override;
     void resolveReferences( const QgsProject &project ) override;
 
   private:
     QgsMapLayerRef mLayerRef; //!< Layer used to extract mesh data from
+    QgsPointCloud3DSymbol *mSymbol = nullptr;
```
Comment (nyalldawson) [inline src/3d/qgspointcloudlayer3drenderer.h:84]:
```suggestion
    std::unique_ptr< QgsPointCloud3DSymbol > mSymbol;
```

File: src/3d/qgspointcloudlayer3drenderer.h [hunk 6]
```diff
@@ -69,12 +70,18 @@ class _3D_EXPORT QgsPointCloudLayer3DRenderer : public QgsAbstract3DRenderer
     QgsPointCloudLayer3DRenderer *clone() const override SIP_FACTORY;
     Qt3DCore::QEntity *createEntity( const Qgs3DMapSettings &map ) const override SIP_SKIP;
 
+    //! Sets 3D symbol associated with the renderer
+    void setSymbol( QgsPointCloud3DSymbol *symbol );
```
Comment (nyalldawson) [inline src/3d/qgspointcloudlayer3drenderer.h:74]:
```suggestion
    /** Sets the 3D \a symbol associated with the renderer.
      * Ownership of \a symbol is transferred to the renderer.
      * \see symbol()
      */
    void setSymbol( QgsPointCloud3DSymbol *symbol SIP_TRANSFER );
```

The usual approach is for "setSymbol" type methods to take ownership of the passed object instead of internally cloning it. It's a little easier to use this way as it can potentially avoid an unwanted double clone if a caller already has an instance of the symbol which they are able to transfer. (and if not, it's trivial for the caller to call `setSymbol( symbol->clone() )`.

File: src/3d/qgspointcloudlayerchunkloader_p.h [hunk 7]
```diff
@@ -121,6 +124,7 @@ class QgsPointCloudLayerChunkLoaderFactory : public QgsChunkLoaderFactory
 
     const Qgs3DMapSettings &mMap;
     QgsPointCloudIndex *mPointCloudIndex;
+    QgsPointCloud3DSymbol *mSymbol = nullptr;
```
Comment (nyalldawson) [inline src/3d/qgspointcloudlayerchunkloader_p.h:127]:
```suggestion
    std::unique_ptr< QgsPointCloud3DSymbol > mSymbol;
```

File: src/3d/qgspointcloudlayerchunkloader_p.h [hunk 8]
```diff
@@ -85,6 +86,8 @@ class QgsPointCloud3DSymbolHandler // : public QgsFeature3DHandler
     // outputs
     PointData outNormal;  //!< Features that are not selected
     // PointData outSelected;  //!< Features that are selected
+
+    QgsPointCloud3DSymbol *mSymbol = nullptr;
```
Comment (nyalldawson) [inline src/3d/qgspointcloudlayerchunkloader_p.h:90]:
```suggestion
    std::unique_ptr< QgsPointCloud3DSymbol > mSymbol;
```

File: src/3d/qgspointcloudlayerchunkloader_p.cpp [hunk 9]
```diff
@@ -274,17 +281,18 @@ Qt3DCore::QEntity *QgsPointCloudLayerChunkLoader::createEntity( Qt3DCore::QEntit
...
+  , mSymbol( symbol )
 {
 }
 
 QgsChunkLoader *QgsPointCloudLayerChunkLoaderFactory::createChunkLoader( QgsChunkNode *node ) const
 {
   QgsChunkNodeId id = node->tileId();
   Q_ASSERT( mPointCloudIndex->hasNode( IndexedPointCloudNode( id.d, id.x, id.y, id.z ) ) );
-  return new QgsPointCloudLayerChunkLoader( this, node );
+  return new QgsPointCloudLayerChunkLoader( this, node, mSymbol );
```
Comment (nyalldawson) [inline src/3d/qgspointcloudlayerchunkloader_p.cpp:295]:
```suggestion
  return new QgsPointCloudLayerChunkLoader( this, node, mSymbol->clone() );
```

File: src/3d/symbols/qgspointcloud3dsymbol.cpp [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgspointcloud3dsymbol.h"
+
+// TODO: For some reason whwn I define function on a .cpp file they don't get included in the qgis_app target
```
Comment (nyalldawson) [inline src/3d/symbols/qgspointcloud3dsymbol.cpp:18]: This would have been caused by the missing _3D_EXPORT macro -- it should be safe to move these back now

File: src/3d/symbols/qgspointcloud3dsymbol.h [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+      Q_UNUSED( context )
+
+      mEnabled = elem.attribute( "enabled", QStringLiteral( "0" ) ).toInt();
+      mPointSize = elem.attribute( "point-size", QStringLiteral( "5.0" ) ).toFloat();
+    }
+
+    /**
+     * Returns whether rendering for this symbol is enabled
+     * \see setIsEnabled( bool enabled )
+     * \since QGIS 3.18
```
Comment (nyalldawson) [inline src/3d/symbols/qgspointcloud3dsymbol.h:72]: We usually don't include redundant \since tags if the member has been introduced in the same version as the class itself (it's effectively redundant because the class itself has the same \since tag)

File: src/app/3d/qgspointcloud3dsymbolwidget.cpp [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    if ( renderer != nullptr )
+    {
+      symbol = const_cast<QgsPointCloud3DSymbol *>( renderer->symbol() );
+      setSymbol( symbol );
+    }
+    else
+    {
+      symbol = new QgsPointCloud3DSymbol;
+      setSymbol( symbol );
+      delete symbol;
```
Comment (nyalldawson) [inline src/app/3d/qgspointcloud3dsymbolwidget.cpp:41]:
```suggestion
      QgsPointCloud3DSymbol symbol;
      setSymbol( &symbol );
```

File: src/app/3d/qgspointcloud3dsymbolwidget.cpp [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+#include "qgspointcloud3dsymbol.h"
+#include "qgspointcloudlayer3drenderer.h"
+
+QgsPointCloud3DSymbolWidget::QgsPointCloud3DSymbolWidget( QgsPointCloudLayer *layer, QWidget *parent )
+  : QWidget( parent )
+  , mLayer( layer )
+{
+  this->setupUi( this );
+
+  if ( layer )
```
Comment (nyalldawson) [inline src/app/3d/qgspointcloud3dsymbolwidget.cpp:28]: Call setLayer instead? The code is the same...

File: src/app/3d/qgspointcloud3dsymbolwidget.cpp [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+void QgsPointCloud3DSymbolWidget::setSymbol( QgsPointCloud3DSymbol *symbol )
+{
+  mPointSizeSpinBox->setValue( symbol->pointSize() );
+}
+
+QgsPointCloud3DSymbol *QgsPointCloud3DSymbolWidget::symbol() const
+{
+  // TODO: fix memory leak
```
Comment (nyalldawson) [inline src/app/3d/qgspointcloud3dsymbolwidget.cpp:76]: This isn't a leak anymore (if you make the suggested change of having the renderer take ownership of the symbol)

File: src/app/3d/qgspointcloudlayer3drendererwidget.cpp [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+  whileBlocking( mChkEnabled )->setChecked( renderer ? renderer->symbol()->isEnabled() : false );
+}
+
+QgsPointCloudLayer3DRenderer *QgsPointCloudLayer3DRendererWidget::renderer()
+{
+  // TODO: use unique_ptr for point cloud symbol
+  QgsPointCloud3DSymbol *sym = mWidgetPointCloudSymbol->symbol();
+  mRenderer.reset( new QgsPointCloudLayer3DRenderer );
+  mRenderer->setSymbol( sym );
+  delete sym;
```
Comment (nyalldawson) [inline src/app/3d/qgspointcloudlayer3drendererwidget.cpp:59]:
```suggestion
```

File: src/app/3d/qgspointcloudlayer3drendererwidget.cpp [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+  mRenderer->setLayer( qobject_cast<QgsPointCloudLayer *>( mLayer ) );
+  return mRenderer.get();
```
Comment (nyalldawson) [inline src/app/3d/qgspointcloudlayer3drendererwidget.cpp:61]: mRenderer doesn't seem to be used anywhere. I'd refine this and make renderer() return a new instance of the renderer (with the caller taking ownership). Then apply() can call `mLayer->setRenderer3D( renderer() )`

File: src/app/3d/qgspointcloudlayer3drendererwidget.cpp [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+
+void QgsPointCloudLayer3DRendererWidget::syncToLayer( QgsMapLayer *layer )
+{
+  mLayer = layer ;
+  QgsPointCloudLayer *pointCloudLayer = qobject_cast<QgsPointCloudLayer *>( layer );
+  mWidgetPointCloudSymbol->setLayer( pointCloudLayer );
+  QgsAbstract3DRenderer *r = layer->renderer3D();
+  if ( r && r->type() == QLatin1String( "pointcloud" ) )
+  {
+    QgsPointCloudLayer3DRenderer *pointCloudRenderer = static_cast<QgsPointCloudLayer3DRenderer *>( r );
```
Comment (nyalldawson) [inline src/app/3d/qgspointcloudlayer3drendererwidget.cpp:90]: This isn't correct -- you're modifying the layer here, whereas syncToLayer should not touch the layer and just update the GUI to reflect the layer's current state

**Comment by NEDJIMAbelgacem** [comment]
> @NEDJIMAbelgacem there's some overlap here with #40019 -- specifically the layer styling and properties dialog work. I suggest we get #40019 merged first as the approach is slightly cleaner in that PR, and then rebase this one to remove the duplicate changes.

Sounds good to me!

**Comment by nyalldawson** [comment]
Can you rebase this now?

File: src/3d/symbols/qgspointcloud3dsymbol.cpp [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+void QgsPointCloud3DSymbol::setIsEnabled( bool enabled )
+{
+  mEnabled = enabled;
+}
+
+void QgsPointCloud3DSymbol::setPointSize( float size )
+{
+  mPointSize = size * 1.0f;
```
Comment (nyalldawson) [inline src/3d/symbols/qgspointcloud3dsymbol.cpp:60]: Why the * 1.0?

File: src/app/pointcloud/qgspointcloudlayerproperties.cpp [hunk 19]
```diff
@@ -104,31 +107,14 @@ QgsPointCloudLayerProperties::QgsPointCloudLayerProperties( QgsPointCloudLayer *
   restoreOptionsBaseUi( title );
 }
 
-void QgsPointCloudLayerProperties::addPropertiesPageFactory( QgsMapLayerConfigWidgetFactory *factory )
```
Comment (nyalldawson) [inline src/app/pointcloud/qgspointcloudlayerproperties.cpp:107]: Can you revert the changes to this file? They aren't needed

File: src/app/qgisapp.cpp [hunk 20]
```diff
@@ -16261,6 +16263,28 @@ void QgisApp::showLayerProperties( QgsMapLayer *mapLayer, const QString &page )
     case QgsMapLayerType::PointCloudLayer:
     {
       QgsPointCloudLayerProperties pointCloudLayerPropertiesDialog( qobject_cast<QgsPointCloudLayer *>( mapLayer ), mMapCanvas, visibleMessageBar(), this );
+
+      //
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:16267]: Remove

File: src/app/qgslayerstylingwidget.cpp [hunk 21]
```diff
@@ -268,6 +256,12 @@ void QgsLayerStylingWidget::setLayer( QgsMapLayer *layer )
 
     case QgsMapLayerType::PointCloudLayer:
     {
+#ifdef HAVE_3D
```
Comment (nyalldawson) [inline src/app/qgslayerstylingwidget.cpp:259]: Remove -- this is handled automatically through the factory

File: src/app/qgslayerstylingwidget.cpp [hunk 22]
```diff
@@ -444,6 +438,10 @@ void QgsLayerStylingWidget::updateCurrentWidgetLayer()
     {
       mMesh3DWidget = widget;
     }
+    else if ( QgsPointCloudLayer3DRendererWidget *widget = qobject_cast<QgsPointCloudLayer3DRendererWidget *>( current ) )
```
Comment (nyalldawson) [inline src/app/qgslayerstylingwidget.cpp:441]: You can remove all the changes to this file -- they are handled automatically through the factory api

File: src/3d/symbols/qgspointcloud3dsymbol.cpp (hunk 18 above)
Comment (NEDJIMAbelgacem) [reply src/3d/symbols/qgspointcloud3dsymbol.cpp:60]: I have no idea when or why I did that. Probably wrote that at 5AM :)

File: src/app/qgslayerstylingwidget.cpp (hunk 22 above)
Comment (NEDJIMAbelgacem) [reply src/app/qgslayerstylingwidget.cpp]: I missed that I need to override supportsStyleDock in QgsPointCloudLayer3DRendererWidgetFactory for the factory to work.
I think it is good now

File: src/3d/symbols/qgspointcloud3dsymbol.cpp [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+  : QgsAbstract3DSymbol()
+{
+
+}
+
+QgsPointCloud3DSymbol::~QgsPointCloud3DSymbol() {  }
+
+QgsAbstract3DSymbol *QgsPointCloud3DSymbol::clone() const
+{
+  // TODO: fix memory leak
```
Comment (nyalldawson) [inline src/3d/symbols/qgspointcloud3dsymbol.cpp:29]: Where's the leak here?

File: src/app/3d/qgspointcloudlayer3drendererwidget.cpp [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+  renderer->setLayer( qobject_cast<QgsPointCloudLayer *>( mLayer ) );
+  return renderer;
+}
+
+void QgsPointCloudLayer3DRendererWidget::apply()
+{
+  QgsPointCloudLayer3DRenderer *r = nullptr;
+  if ( mChkEnabled->isChecked() )
+  {
+    r = renderer()->clone();
```
Comment (nyalldawson) [inline src/app/3d/qgspointcloudlayer3drendererwidget.cpp:66]:
```suggestion
    r = renderer();
```

renderer() is already a factory

File: src/app/pointcloud/qgspointcloudlayerproperties.h [hunk 25]
```diff
@@ -71,6 +72,9 @@ class QgsPointCloudLayerProperties : public QgsOptionsDialogBase, private Ui::Qg
     QgsMapCanvas *mMapCanvas = nullptr;
     QgsMetadataWidget *mMetadataWidget = nullptr;
 
+    //! A list of additional pages provided by plugins
+    QList<QgsMapLayerConfigWidget *> mLayerPropertiesPages;
```
Comment (nyalldawson) [inline src/app/pointcloud/qgspointcloudlayerproperties.h:76]: Remove

File: src/app/qgisapp.cpp [hunk 26]
```diff
@@ -16261,6 +16263,12 @@ void QgisApp::showLayerProperties( QgsMapLayer *mapLayer, const QString &page )
     case QgsMapLayerType::PointCloudLayer:
     {
       QgsPointCloudLayerProperties pointCloudLayerPropertiesDialog( qobject_cast<QgsPointCloudLayer *>( mapLayer ), mMapCanvas, visibleMessageBar(), this );
+
+      if ( !page.isEmpty() )
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:16267]: Duplicate of code which starts at line 16272?

File: src/app/qgslayerstylingwidget.cpp [hunk 27]
```diff
@@ -812,7 +790,7 @@ bool QgsLayerStyleManagerWidgetFactory::supportsLayer( QgsMapLayer *layer ) cons
       return false;  // TODO
 
     case QgsMapLayerType::PointCloudLayer:
-      return false;  // TODO
```
Comment (nyalldawson) [inline src/app/qgslayerstylingwidget.cpp:815]: This change isn't used anywhere -- QgsLayerStyleManagerWidgetFactory isn't inherited by any of the point cloud QgsMapLayerConfigWidgetFactory subclasses.

I'd suggest collapsing the point cloud (and vector tile layer) cases into the final plugin/annotation layer case and removing the "TODO" notes, since there's no TODO required here.

**Comment by nyalldawson** [comment]
LGTM, thanks! Not sure if @PeterPetrik / @wonder-sk also want to review?

**Review by PeterPetrik — APPROVED** [review]
Looks good to me
